File: Nejman.Newtone.Mobile/Services/ContextMenuBuilder.cs
using Nejman.Newtone.Core;
using Nejman.Newtone.Core.Data;
using Nejman.Newtone.Core.Implementations;
using Nejman.Newtone.Core.Localization;
using Nejman.Newtone.Mobile.Implementations;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Xamarin.Forms;

namespace Nejman.Newtone.Mobile.Services
{
    public static class ContextMenuBuilder
    {
        #region Properties
        private static string CurrentModelInfo { get; set; }
        #endregion
        #region Public Methods
        public static void BuildForTrack(View sender, string modelInfo)
        {
            CurrentModelInfo = modelInfo;

            string[] elems = CurrentModelInfo.Split(new[] { CoreGlobal.SEPARATOR }, StringSplitOptions.None);
            string filePath = elems[0];
            string playlistName = elems[1];

            List<string> menuItems = new List<string>() { Localization.TrackMenuEdit, Localization.TrackMenuPlaylist, Localization.TrackMenuQueue, Localization.TrackMenuDelete };

           ContextMenuBuilderImplementation.Current.BuildForTrack(sender, modelInfo, filePath, playlistName, menuItems, TrackAction);
        }

        public static void BuildForPlaylist(View sender, string playlistName)
        {
            List<string> elements = new List<string>() { Localization.PlaylistPlay, Localization.TrackMenuPlaylist, Localization.TrackMenuQueue, Localization.ChangeName, Localization.TrackMenuDelete };

            ContextMenuBuilderImplementation.Current.BuildForPlaylist(sender, playlistName, elements, PlaylistAction);
        }

        public static void BuildForArtist(View sender, string artistName)
        {
            List<string> elements = new List<string>() { Localization.PlaylistPlay, Localization.TrackMenuPlaylist, Localization.TrackMenuQueue };

            ContextMenuBuilderImplementation.Current.BuildForArtist(sender, artistName, elements, ArtistAction);
        }

        public static void BuildForSearchResult(View sender, string modelInfo)
        {
            List<string> elements = new List<string>() { Localization.Download };

            ContextMenuBuilderImplementation.Current.BuildForSearchResult(sender, modelInfo, elements, SearchResultAction);
        }
        #endregion
        #region Private Methods
        private static async Task TrackAction(string filePath, string item, string playlistName)
        {
            if (filePath.Length == 11)
            {
                SnackbarImplementation.Current.Show(Localization.SnackFileExists);
                return;
            }

            if (item == Localization.TrackMenuEdit)
            {
                await TracksAction.Edit(filePath);
            }
            else if (item == Localization.TrackMenuPlaylist)
            {
                await PlaylistsAction.Add(filePath);
            }
            else if (item == Localization.TrackMenuDelete)
            {
                if(playlistName != "")
                {
                    await PlaylistsAction.Remove(playlistName, filePath);
                }
                else
                {
                    await TracksAction.Remove(filePath);
                }
            }
            else if (item == Localization.TrackMenuQueue)
            {
                QueueAction.Add(filePath);
                SnackbarImplementation.Current.Show(Localization.SnackQueue);
            }
        }
        private static async Task PlaylistAction(View sender, string playlistName, string item)
        {
            if (item == Localization.PlaylistPlay)
            {
                var playlist = PlaylistsAction.GetPlaylist(playlistName);

                if(playlist.Count > 0)
                {
                    await CoreGlobal.MediaPlayer.LoadPlaylist(playlist);
                }
            }
            else if (item == Localization.TrackMenuPlaylist)
            {
                var playlist = PlaylistsAction.GetPlaylist(playlistName);
                await PlaylistsAction.Add(playlist.Select(source => source.Path).ToList());
            }
            if (item == Localization.ChangeName)
            {
                await PlaylistsAction.ChangeName(playlistName);
            }
            else if (item == Localization.TrackMenuDelete)
            {
                await PlaylistsAction.Remove(playlistName);
            }
            else if (item == Localization.TrackMenuQueue)
            {
                var playlist = PlaylistsAction.GetPlaylist(playlistName);
                QueueAction.Add(playlist.Select(source => source.Path).ToList());
                SnackbarImplementation.Current.Show(Localization.SnackQueue);
            }
        }
        private static async Task ArtistAction(View sender, string artistName, string item)
        {
            if (item == Localization.PlaylistPlay)
            {
                var artist = ArtistsAction.GetArtist(artistName);

                if (artist.Count > 0)
                {
                    await CoreGlobal.MediaPlayer.LoadPlaylist(artist);
                }
            }
            else if (item == Localization.TrackMenuPlaylist)
            {
                var artist = ArtistsAction.GetArtist(artistName);
                await PlaylistsAction.Add(artist.Select(source => source.Path).ToList());
            }
            else if (item == Localization.TrackMenuQueue)
            {
                var artist = ArtistsAction.GetArtist(artistName);
                QueueAction.Add(artist.Select(source => source.Path).ToList());
                SnackbarImplementation.Current.Show(Localization.SnackQueue);
            }
        }
#pragma warning disable CS1998 // Metoda asynchroniczna nie zawiera operatorów „await” i zostanie uruchomiona synchronicznie
        private static async Task SearchResultAction(View sender, string tag, string item)
#pragma warning restore CS1998 // Metoda asynchroniczna nie zawiera operatorów „await” i zostanie uruchomiona synchronicznie
        {
            if (item == Localization.Download)
            {
                await DownloadAction.Add(tag);
            }
        }
        #endregion
    }
}

File: app/src/main/java/com/nejman/nsec/music_player/ui/ContextMenuBuilder.java
package com.nejman.nsec.music_player.ui;

import android.view.View;
import android.widget.PopupMenu;
import android.widget.Toast;

import com.nejman.nsec.music_player.Global;
import com.nejman.nsec.music_player.MainActivity;
import com.nejman.nsec.music_player.R;
import com.nejman.nsec.music_player.core.DataContainer;
import com.nejman.nsec.music_player.core.data.PlaylistsAction;
import com.nejman.nsec.music_player.core.data.QueueAction;
import com.nejman.nsec.music_player.core.data.TracksAction;
import com.nejman.nsec.music_player.core.models.ArtistModel;
import com.nejman.nsec.music_player.core.models.PlaylistModel;
import com.nejman.nsec.music_player.media.NewtoneMediaPlayer;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class ContextMenuBuilder {
    public static void buildForSearchResult(View sender, String modelInfo) {
        List<String> elements = new ArrayList<>();
        elements.add(MainActivity.instance.getString(R.string.download));
        build(sender, elements, item -> {
            if (item.equals(MainActivity.instance.getString(R.string.download))) {
                DataContainer.getInstance().getDownloads().add(modelInfo);
            }
        });
    }

    public static void buildForPlaylist(View sender, String playlistName) {
        List<String> elements = new ArrayList<>();
        elements.add(MainActivity.getResString(R.string.playlist_play));
        elements.add(MainActivity.getResString(R.string.track_menu_playlist));
        elements.add(MainActivity.getResString(R.string.track_menu_queue));
        elements.add(MainActivity.getResString(R.string.change_name));
        elements.add(MainActivity.getResString(R.string.track_menu_delete));

        build(sender, elements, item -> {
            if (item.equals(MainActivity.getResString(R.string.playlist_play))) {
                PlaylistModel model = DataContainer.getInstance().getPlaylists().get(playlistName);

                if (model.items.size() > 0) {
                    NewtoneMediaPlayer.getInstance().loadPlaylist(model.getSources(), 0);
                }
            } else if (item.equals(MainActivity.getResString(R.string.track_menu_playlist))) {
                PlaylistModel model = DataContainer.getInstance().getPlaylists().get(playlistName);
                PlaylistsAction.add(model.items);
            } else if (item.equals(MainActivity.getResString(R.string.change_name))) {
                PlaylistsAction.changeName(playlistName);
            } else if (item.equals(MainActivity.getResString(R.string.track_menu_delete))) {
                PlaylistsAction.remove(playlistName);
            } else if (item.equals(MainActivity.getResString(R.string.track_menu_queue))) {
                PlaylistModel model = DataContainer.getInstance().getPlaylists().get(playlistName);
                QueueAction.add(model.items);
                Toast.makeText(MainActivity.instance, MainActivity.getResString(R.string.snack_queue), Toast.LENGTH_SHORT).show();
            }
        });
    }

    public static void buildForTrack(View sender, String modelInfo) {
        String[] elems = modelInfo.split(Global.separator);
        String filePath = elems[0];
        String playlistName = elems.length == 2 ? elems[1] : "";
        List<String> elements = new ArrayList<>();
        elements.add(MainActivity.getResString(R.string.track_menu_edit));
        elements.add(MainActivity.getResString(R.string.track_menu_playlist));
        elements.add(MainActivity.getResString(R.string.track_menu_queue));
        elements.add(MainActivity.getResString(R.string.track_menu_delete));
        build(sender, elements, item -> {
            if (filePath.length() == 11) {
                MainActivity.toast(R.string.snack_file_exists);
                return;
            }

            if (item.equals(MainActivity.getResString(R.string.track_menu_edit))) {
                TracksAction.edit(DataContainer.getInstance().getMediaSources().get(filePath));
            } else if (item.equals(MainActivity.getResString(R.string.track_menu_playlist))) {
                PlaylistsAction.add(filePath);
            } else if (item.equals(MainActivity.getResString(R.string.track_menu_delete))) {
                if (playlistName.equals("")) {
                    TracksAction.remove(filePath);
                } else {
                    PlaylistsAction.remove(playlistName, filePath);
                }
            } else if (item.equals(MainActivity.getResString(R.string.track_menu_queue))) {
                QueueAction.add(filePath);
                Toast.makeText(MainActivity.instance, MainActivity.getResString(R.string.snack_queue), Toast.LENGTH_SHORT).show();
            }
        });
    }

    public static void buildForArtist(View sender, String artistName) {
        List<String> elements = new ArrayList<>();
        elements.add(MainActivity.getResString(R.string.playlist_play));
        elements.add(MainActivity.getResString(R.string.track_menu_playlist));
        elements.add(MainActivity.getResString(R.string.track_menu_queue));

        build(sender, elements, item -> {
            if (item.equals(MainActivity.getResString(R.string.playlist_play))) {
                ArtistModel model = DataContainer.getInstance().getArtists().get(artistName);

                if (model.items.size() > 0) {
                    NewtoneMediaPlayer.getInstance().loadPlaylist(model.getSources(), 0);
                }
            } else if (item.equals(MainActivity.getResString(R.string.track_menu_playlist))) {
                ArtistModel model = DataContainer.getInstance().getArtists().get(artistName);
                PlaylistsAction.add(model.items);
            } else if (item.equals(MainActivity.getResString(R.string.track_menu_queue))) {
                ArtistModel model = DataContainer.getInstance().getArtists().get(artistName);
                QueueAction.add(model.items);
                Toast.makeText(MainActivity.instance, MainActivity.getResString(R.string.snack_queue), Toast.LENGTH_SHORT).show();
            }
        });
    }

    private static void build(View anchor, List<String> elements, Consumer<? super String> action) {
        PopupMenu menu = new PopupMenu(MainActivity.instance.getApplicationContext(), anchor);

        for (String element : elements) {
            menu.getMenu().add(element);
        }
        menu.setOnMenuItemClickListener(item -> {
            action.accept(item.getTitle().toString());
            return true;
        });
        menu.show();
    }
}
